### ISSUE TITLE
IDE crashed when trying to do edit->preferences

### ISSUE THREAD

**Issue by richardwzp** [coq/coq#16136]
```
-------------------------------------
Translated Report (Full Report Below)
-------------------------------------

Process:               coqide.exe [7799]
Path:                  /Applications/Coq-Platform~8.15~2022.04.app/Contents/Resources/bin/coqide.exe
Identifier:            coqide.exe
Version:               ???
Code Type:             X86-64 (Native)
Parent Process:        launchd [1]
User ID:               503

Date/Time:             2022-05-22 17:44:01.5214 -0400
OS Version:            macOS 12.3.1 (21E258)
Report Version:        12
Bridge OS Version:     6.5 (19P5071)
Anonymous UUID:        D3DFB2CA-165E-DE9E-8BC6-33029E6CAF79

Sleep/Wake UUID:       01F49602-791E-4A05-A683-64D2A6D85425

Time Awake Since Boot: 7200 seconds
Time Since Wake:       6040 seconds

System Integrity Protection: enabled

Crashed Thread:        0  Dispatch queue: com.apple.main-thread

Exception Type:        EXC_BAD_ACCESS (SIGSEGV)
Exception Codes:       KERN_INVALID_ADDRESS at 0x0000000000000008
Exception Codes:       0x0000000000000001, 0x0000000000000008
Exception Note:        EXC_CORPSE_NOTIFY

Termination Reason:    Namespace SIGNAL, Code 11 Segmentation fault: 11
Terminating Process:   exc handler [7799]

VM Region Info: 0x8 is not in any region.  Bytes before following region: 4511461368
      REGION TYPE                    START - END         [ VSIZE] PRT/MAX SHRMOD  REGION DETAIL
      UNUSED SPACE AT START
--->  
      __TEXT                      10ce77000-10d1e7000    [ 3520K] r-x/r-x SM=COW  ...in/coqide.exe
```
I was on BigSur with the january version, and preference just crashes the ide. I updated to Monterey, and is using `coqIDE 8.15`, and this issue seems to still persist.

**Comment by xaviripo** [comment]
> OK, and what happens if you delete this file?

Same behavior. Also the program does not recreate the file.

**Comment by jfehrle** [comment]
Your `coqiderc` looks unexceptional.  If this was a problem in the OCaml code that parses the preferences, I'd expect we'd see an OCaml exception rather than a segfault.  I think it's more likely a GTK bug.  Would you run `coqide` with the `-debug` option?  Note the log file name in the messages panel when the window comes up, then make it crash and send in the log file.  There's a small chance that would give a clue to what's happening.

**Comment by orilahav** [comment]
Is this the same bug report as [#15486](https://github.com/coq/coq/issues/15486)?

**Comment by xaviripo** [comment]
I ran `coqide -debug 2> coqide.log` as the messages panel said it was just logging to `stderr`.

coqide.log:

```
(Spawn  ,49229) EXEC: coqidetop.opt -main-channel stdfds --xml_format=Ppcmds -async-proofs on
(Spawn  ,49229) PID 49232
[DEBUG] Start eval_call Init None
[DEBUG] End eval_call
[DEBUG] End of Coqide.main

(coqide:49229): Gtk-WARNING **: 11:04:22.453: Drawing a gadget with negative dimensions. Did you forget to allocate a size? (node header owner GtkNotebook)

(coqide:49229): Gtk-WARNING **: 11:04:22.453: Drawing a gadget with negative dimensions. Did you forget to allocate a size? (node tabs owner GtkNotebook)

(coqide:49229): Gtk-WARNING **: 11:04:22.453: Drawing a gadget with negative dimensions. Did you forget to allocate a size? (node tab owner GtkNotebook)
[DEBUG] Handling coqtop answer
[DEBUG] Start eval_call SetOptions [(["Diffs"],off);(["Printing";"Coercions"],false);(["Printing";"Matching"],true);(["Printing";"Notations"],true);(["Printing";"Existential";"Instances"],false);(["Printing";"Unfocused"],false);(["Printing";"Goal";"Names"],false);(["Printing";"Parentheses"],false);(["Printing";"Implicit"],false);(["Printing";"All"],false);(["Printing";"Universes"],false)]
[DEBUG] End eval_call
[DEBUG] Handling coqtop answer
got key '' (65511, )
got key ',' (44, MOD2 META)

```

Besides this, when trying to open settings, this time I got a pop-up window with message:

```
Coqide internal error: file ../gtk/gtkcssenumvalue.c: line 529 ( gtk css_font weight_value new): should not be reached 
```

Before crashing.

**Comment by MollyDream** [comment]
I have encountered the same issue on coqide 8.13.2 and MacOS Monterey 12.0.1. The -debug log:
```
coqide -debug
(Spawn  ,90557) EXEC: coqidetop.opt -main-channel stdfds --xml_format=Ppcmds -async-proofs on -debug
(Spawn  ,90557) PID 90560
[DEBUG] Start eval_call Init None
[DEBUG] End eval_call
[DEBUG] End of Coqide.main

(coqide:90557): Gtk-WARNING **: 16:05:17.338: Drawing a gadget with negative dimensions. Did you forget to allocate a size? (node header owner GtkNotebook)

(coqide:90557): Gtk-WARNING **: 16:05:17.338: Drawing a gadget with negative dimensions. Did you forget to allocate a size? (node tabs owner GtkNotebook)

(coqide:90557): Gtk-WARNING **: 16:05:17.338: Drawing a gadget with negative dimensions. Did you forget to allocate a size? (node tab owner GtkNotebook)
(Spawn  ,90560) EXEC: coqproofworker.opt -main-channel stdfds -debug -worker-id proofworker:0 -async-proofs-worker-priority high
(Spawn  ,90560) PID 90561
90560:master:2] waiting for a task
[pid 90560] <-- Init None
[pid 90560] --> GOOD 1
[DEBUG] Handling coqtop answer
[DEBUG] Start eval_call SetOptions [(["Diffs"],off);(["Printing";"Coercions"],false);(["Printing";"Matching"],true);(["Printing";"Notations"],true);(["Printing";"Existential";"Instances"],false);(["Printing";"Unfocused"],false);(["Printing";"Goal";"Names"],false);(["Printing";"Parentheses"],false);(["Printing";"Implicit"],false);(["Printing";"All"],false);(["Printing";"Universes"],false)]
[DEBUG] End eval_call
[pid 90560] <-- SetOptions [(["Diffs"],off);(["Printing";"Coercions"],false);(["Printing";"Matching"],true);(["Printing";"Notations"],true);(["Printing";"Existential";"Instances"],false);(["Printing";"Unfocused"],false);(["Printing";"Goal";"Names"],false);(["Printing";"Parentheses"],false);(["Printing";"Implicit"],false);(["Printing";"All"],false);(["Printing";"Universes"],false)]
[pid 90560] --> GOOD
[DEBUG] Handling coqtop answer
[pid 90560] End of input, exiting gracefully.
zsh: segmentation fault  coqide -debug
90561:proofworker:0:0] connection lost
```
and the pop-up error window 
`Coqide internal error: file ../gtk/gtkcssenumvalue.c: line 529 ( gtk css_font weight_value new): should not be reached`

I have previously had seg fault whenever I run a Coq file, and I got it solved by upgrading to lablgtk3 3.1.2 following the advice [here](https://github.com/coq/coq/issues/14352#issuecomment-1006629498). Now the CoqIDE runs, but very slowly, so I want to configure the resolution, and that is how I ran into this issue.

**Comment by MSoegtropIMC** [comment]
@MollyDream : did you also try to open preferences, or did it crash in a different situation?

**Comment by dnaumann** [comment]
CoqIDE crashes whenever I open preferences.  On 2019 Intel i9 MacBook and Coq Platform 2022.04.1

**Comment by orilahav** [comment]
This bug is quite systematic, and open here for some months... Is there any way users can contribute to analyzing the source of this problem?

**Comment by Alizter** [comment]
@coq/coqide-maintainers What should we do about this? I've tried looking for a GTK bug but I couldn't find one. Do we have some way of making a MacOS-specific workaround?

**Comment by MSoegtropIMC** [comment]
@dnaumann, @orilahav : can you please add the information here if you used Coq Platform installers (DMG) or Coq Platform compile from sources scripts - preferably by editing your above posts.

**Comment by orilahav** [comment]
I installed it via opam

**Comment by jfehrle** [comment]
@Alizter Here are some things I would try if I had a MacOS development environment.  From the descriptions provided, it's most likely failing when GTK tries to display the dialog.  But it could be something else.

1. Localize the failure by adding print statements to CoqIDE, eg. just before `let x = edit ~apply "Customizations" ~parent cmds in` in `preferences.ml`, then at various points in `Configwin_ihm.edit` which it calls.  You should be able to localize it to a single line.

2. Remove various subpanels in `cmds` in `preferences.ml` to see if it's related to a specific panel (e.g. remove all but one to start).  If it's related to a single panel, then try removing items within the panel to localize the problem.

If you like, we could do a joint debugging/brainstorming session over Zoom sometime this weekend--let me know if/when you'd like to do that.  But I suggest trying my suggestions above on your own first.

**Comment by jfehrle** [comment]
Have you verified you can reproduce the problem if you build from the dev tree?

Also, can you figure out what is the last good version of CoqIDE (not clear in #15486)?    Note that it could be due to a difference in the lablgtk, in which case you may have to try various older versions of these libraries.  (Much less likely, it may be caused by a change to the OCaml compiler version or to dune.)

```
opam list
# Packages matching: installed
# Name                  # Installed # Synopsis
lablgtk3                3.1.2       OCaml interface to GTK+3
lablgtk3-sourceview3    3.1.2       OCaml interface to GTK+ gtksourceview library
```

**Comment by jfehrle** [comment]
PS I've assumed this always fails for a given (Coq version, MacOS version) and thus the problem should be readily reproducible.

**Comment by Zimmi48** [comment]
> Also, can you figure out what is the last good version of CoqIDE (not clear in #15486)? I see @MollyDream reports the problem appears in Coq 8.13.2 (above).

It was also mentioned by @amosunov in https://github.com/coq/coq/issues/15486#issuecomment-1156786466 that the DMG installer for 8.13.0 does not have the bug. Since I don't think there was any change in the Coq sources that could explain this difference between 8.13.0 and 8.13.2, it seems more likely that it is a change in an external dependency (like GTK) that is linked in the DMG image. Perhaps we could inspect the DMG images to figure the differences.
The most noticeable difference between the 8.13.0 DMG and the 8.13.2 DMG is that the former was generated from the Coq repository using our old DMG creation pipeline, whereas the latter was generated from the Coq Platform repository. Therefore, there are big differences between the two.

To better understand the source of the problem, it would be therefore useful to:
- confirm that the [8.13.0 DMG](https://github.com/coq/coq/releases/download/V8.13.0/coq-8.13.0-installer-macos.dmg) does not have the bug
- test if the [8.13.1 DMG](https://github.com/coq/coq/releases/download/V8.13.1/coq-8.13.1-installer-macos.dmg) (last before Coq Platform) also does not have the bug
- since it was confirmed that the bug also happens when building with opam, test whether building 8.13.0 from sources also leads to the bug (on a machine where the bug was reproduced when building with opam).

This would help confirm that the bug is independent on changes in the Coq sources and that it is related to a bad combination of versions of some dependencies (e.g., GTK) and recent versions of macOS...

**Comment by orilahav** [comment]
I tried some DMGs:
8.13.0 DMG works
8.13.1 DMG (last before Coq Platform) works
8.13.2 DMG (via Coq Platform): strange error msg: "Editing preferences failed (Not_found)"
8.14.1 DMG (via Coq Platform): crashes as described in this issue
8.15.2 DMG (via Coq Platform): crashes as described in this issue

**Comment by Zimmi48** [comment]
@orilahav Thanks! Would you mind also testing 8.13.1 with opam?

**Comment by orilahav** [comment]
updated my comment above. I don't think I can check earlier version, bcs of minimal ocmal version on M1 IIUC.

**Comment by Zimmi48** [comment]
So it seems that the first error ("Editing preferences failed") is not due to a change in the Coq sources but to a change in the environment, most likely which macOS version was used to build CoqIDE. For the record, until 8.13.1, the DMG image was built in an Azure pipeline running macOS 10.14 (https://github.com/coq/coq/blob/V8.13.1/azure-pipelines.yml#L53) whereas the Coq Platform DMG image was built in GitHub Actions using `macos-latest`, which (from what I understand from parsing https://github.blog/?s=macos-latest) meant macOS 10.15 at the time of building the first Coq Platform DMG image. Note that this was not the only change in the environment between these two builds.

Besides, the "Editing preferences failed" error message comes from https://github.com/coq/coq/blob/06b78fba967fd10b7ef18e977935db3e305dee14/ide/coqide/coqide.ml#L338. We can imagine that something happened in macOS 11 that made it impossible to save the preference file and something changed in the Coq sources after 8.13.2 that turned a `Not_found` exception into a segfault.

**Comment by orilahav** [comment]
isn't it the following line?
https://github.com/coq/coq/blob/06b78fba967fd10b7ef18e977935db3e305dee14/ide/coqide/coqide.ml#L1459

(just by the text of the message)

**Comment by jfehrle** [comment]
> We can imagine that something happened in macOS 11 that made it impossible to save the preference file

It's unlikely that saving the preference file is the cause; `flash_info` is called for any failure in the callback, which handles all the processing for the dialog. IIUC the preferences dialog isn't even get drawn.  Not much point in saving the preference file before the user has made any change.

@orilahav I expect there's a way to cause MacOS to generate a core file after the segment fault.  If so, you might be able to run a debugger such as `gdb` on the core file and get a stack trace.  That might give some helpful clues, particularly if the binary file includes debug symbols.

**Comment by MSoegtropIMC** [comment]
@Zimmi48 : please note that @orilahav reported [above](https://github.com/coq/coq/issues/16136#issuecomment-1242321280) that CoqIDE installed via opam doesn't work either.

**Comment by Zimmi48** [comment]
> @Zimmi48 : please note that @orilahav reported [above](https://github.com/coq/coq/issues/16136#issuecomment-1242321280) that CoqIDE installed via opam doesn't work either.

@MSoegtropIMC Yes, I'm well aware. This might not be clear in my comment, but this is from the inconsistency in the versions that introduce the first error between the DMG and opam that I deduce that the first cause is a change in the environment.

**Comment by MSoegtropIMC** [comment]
@Zimmi48 : I can't quite follow your logic - if it also happens with opam, how can it be an issue with the installer?

Btw.: we should check when @gares introduced the reading of a config file - afair it was in this time frame. It is the only change in CoqIDE I am aware of which might have something to do with it.

Otherwise it might also be a change in GTK versions - I don't pin these in Coq Platform (maybe I should).

**Comment by Zimmi48** [comment]
@MSoegtropIMC I never said it was an issue with the installer, I said the cause of the first issue (the "Editing preference failed") was a change in the environment. My reason for deducing this is that the first version @orilahav observed this error message differs between opam and the installer. opam-compiled Coq has obviously a similar build environment (non-OCaml libs and the macOS version), whereas the installers did not have the same build environment (macOS 10.14 vs macOS 10.15). I'm saying this to help understand the cause of the issue, not to say that it is the installer that should be fixed!

Thus, I don't think it is a change in the Coq sources that introduced the first error, however it is probably a change in the Coq sources that turned this error (uncaught exception) into a segfault. But this change in the Coq sources could also be a change in the build system (the transition to Dune for OCaml libs happened in 8.14).

**Comment by herbelin** [comment]
Hi, was the problem reproducible with a compilation from sources? Based on the presence of font-related functions close to the crash and on a reference to fonts in a seemingly related report ([gnote](https://gitlab.gnome.org/GNOME/gnote/-/issues/22)), I wonder whether the call to `set_font_name text_font#get` in `Preferences.configure` could not be a cause of the  `Not_found`. At least, this field is dependent on `Coq_config.gtk_platform` (`X11` or `QUARTZ`), while the presence or attributes of the font itself might plausibly be dependent on the OS.

What is the value of `gtk_platform` depending on the platform, homebrew, and opam installation?

**Comment by Zimmi48** [comment]
>Hi, was the problem reproducible with a compilation from sources?

You mean directly from sources without opam?

**Comment by MSoegtropIMC** [comment]
@herbelin : regarding the X11 / quartz question : on MacPorts I require Quartz via opam, on Homebrew not, simply because I don't know how. Please also note that the "gtksourceview3 +quartz" syntax only works by luck, but the opam team is aware that I am doing this.
```
opam show conf-gtksourceview3 --raw
:
depexts: [
:
  ["gtksourceview3" "libxml2"] {os = "macos" & os-distribution = "homebrew"}
  ["gtksourceview3 +quartz" "libxml2"] {os = "macos" & os-distribution = "macports"}
:
]
```
I do not install any fonts, so what fonts are there depends on the system and font translation. This might depend on if Coq is installed or not - some GTK package might install some fonts, but I definitely don't copy any fonts into the installer (only shared libraries, icons and cursors and a handful of config files).

**Comment by herbelin** [comment]
@Zimmi48, @MSoegtropIMC, thanks for the answers. I could reproduce the segfault (even with Coq say 8.11 or an old MacOS). The critical line is [`box#pack ~expand:true w#coerce`](https://github.com/coq/coq/blob/dd58e6bc4981096571c66788c92d18ee36dfd4d3/ide/coqide/preferences.ml#L794) activating the (deprecated) font selection. I don't know what caused the regression (I'm using lablgtk3 3.1.2 and gtk3 2.24.32_3) nor am I able to explain what happens.

**Comment by MSoegtropIMC** [comment]
@herbelin : does the edit of your post mean it works with 8.11, or didn't you test it as yet?

**Comment by herbelin** [comment]
With versions before 8.14, it displays "Editing preferences failed (Not_found)" and does not segfault. Will soon report result of bisecting.

**Comment by MSoegtropIMC** [comment]
OK, but this is not much better then. I think we should concentrate on fixing it.

The report might not be a result of a regression, but simply an effect of a larger number of users. The number of CoqIDE users which come across the preferences dialog on a system on which this does not work might have been not that large in the past. The Mac installers and Mac fixes in opam around GTK I did are not that old (8.13 or so).

**Comment by herbelin** [comment]
I don't know why I can reproduce a failure of the preference menu on my MacOS 10.3.6 setting (compiled from the git archive) while beforehand the preference menu was working for me.

In any case, it is from commit 32cc7cb03f879d089e226958ab8e89cbcde79d10 (#14060, ability to use MacOS `<Cmd>` as a modifier) that the failure of the preference menu changed from the Not_found message to a segfault.

More precisely, the line that lead to observing a segfault is the [addition](https://github.com/coq/coq/blob/dd58e6bc4981096571c66788c92d18ee36dfd4d3/ide/coqide/configwin_ihm.ml#L48) of a printer for the `META` key (i.e. <cmd> on MacOS). Without the printer, a Not_found was raised when trying to print `META` and this was stopping the display of the preferences menu without going to the step (apparently in font selection, see previous comment) that causes a segfault.

So, shortly, on Big Sur, Monterey and for some unknown reason also on my old Sierra setup, the preference dialog box fails to open. Before 8.14, this was observed through a failure to print the `META` key, but then, by supporting `META`, PR #14060 lead to reveal a second issue.

About this second issue, I don't know what to do. "Font selection" is told to be deprecated and to be replaced by "font chooser" but I did not see the latter supported by lablgtk. So, I don't see what to do else at the current time than radically removing the font selection tab from the preference box.

In particular, @MSoegtropIMC, I would not consider it as a regression In Coq (even if observed differently after and before 8.14). I would rather say that gtk+ behaves differently depending on specific circonstances. In my case, a posible reason is that my laptop screen is broken and I'm using an external monitor (in particular, I already observed that coqide had some gtk+ rendering issues on external monitors, like cursor not showing when hovering over menu items, or so), but that is only a guess.

Also, since I can reproduce by compiling from the git archive, I don't think that the installers have to be blamed in any way.

**Comment by MSoegtropIMC** [comment]
@herbelin : it might be that the ability to show the preferences dialog depends on the existence / contents of the preferences files. I would check if you can reproduce if you delete these files.

**Comment by herbelin** [comment]
@MSoegtropIMC: good point, it indeed matters. If I remove the preference files, it segfaults even when it would otherwise display Not_found (e.g. with Coq 8.13). Investigating further, it seems that to encounter the Not_found message, the coqiderc `modifiers_valid` field has to include the `<Primary>` key. In particular, the segfault can be triggered on Coq < 8.14 by removing `<Primary>` from the list of `modifiers_valid` in the coqiderc file.

At least, this says that PR #14060 and the difference of behavior between 8.13 and 8.14 is incidental. E.g., running 8.13 with its own preference file (rather than with a preference file produced by 8.14 and containing `<Primary>`), leads to a segfault also in 8.13.

**Comment by MSoegtropIMC** [comment]
@herbelin : So there are two issues? One with the primary modifier which results in some versions of Coq in a `Not_found` and this hiding the seg fault. And then there is the font issue which results in the seg fault, unless one gets the `Not_found` before?

**Comment by herbelin** [comment]
Yes, as a summary, there are two issues, the second one (at least) being only on Big Sur and Monterey iiuc (with an exception for me). MacPorts vs homebrew or M1 vs Intel or Quartz vs X11 or opam vs platform vs manual a priori do not matter (though a confirmation from someone compiling from sources using Big Sur or Monterey could be useful). Unclear to me is whether ocaml or lablgtk or gtk versions matter (I think I tried once with ocaml 4.06.1 and lablgtk 3.1.1 and it fails, but this should be double-checked).

Afaiu, the Not_found issue only occurs when using Coq < 8.14 with a preference file generated with Coq >= 8.14, in which case it hides the second issue.

The second (and main) issue is a crash of `_gtk_css_value_ref` after a `Coqide internal error: file ../gtk/gtkcssenumvalue.c: line 529 ( gtk css_font weight_value new): should not be reached` debug message. It can be prevented by recompiling without the drawing of the `font` tab of the preference dialog box.

**Comment by Zimmi48** [comment]
It could be worth fixing this the easy way (disable the font preference on macOS) and release a new version in time for 2022.09.0?

**Comment by MSoegtropIMC** [comment]
It won't make it for the beta, but the main release is in about 2 1/2 weeks.

**Comment by Zimmi48** [comment]
cc @ppedrot and @SkySkimmer

**Comment by SkySkimmer** [comment]
?

**Comment by Zimmi48** [comment]
You are listed as a co-RM for 8.16, so I'm bringing this to your attention: https://github.com/coq/coq/issues/16136#issuecomment-1262116228

**Comment by ppedrot** [comment]
I can even write the fix myself. What's the most correct path? Should we still have a degraded font preference with just a text field and parse this as a font name through the GTK API? When should we go into degraded mode? Systematically when run under MacOS?

**Comment by ppedrot** [comment]
Given that the font dialog segfault doesn't seem more common in the wild, we might be doing something wrong. Out of curiosity, can somebody check whether the following diff makes any change?
```diff
--- a/ide/coqide/preferences.ml
+++ b/ide/coqide/preferences.ml
@@ -788,7 +788,7 @@ let configure ?(apply=(fun () -> ())) parent =
 
   let config_font =
     let box = GPack.hbox () in
-    let w = GMisc.font_selection () in
+    let w = GMisc.font_selection ~font_name:text_font#get () in
     w#set_preview_text
       "Goal (∃n : nat, n ≤ 0)∧(∀x,y,z, x∈y⋃z↔x∈y∨x∈z).";
     box#pack ~expand:true w#coerce;
```
With a bit of luck this does not take the same code path as the one where we leave the default font.

In the meantime I'll write a proper workaround.

**Comment by herbelin** [comment]
> can somebody check whether the following diff makes any change?

Still a segfault (but maybe this is what your "EDIT" anticipated).

> we might be doing something wrong.

Possibly. I've seen e.g. segfault on _gtk_css_value_ref reported on two places: [epiphany](https://bugzilla.gnome.org/show_bug.cgi?id=747422) (weak pointer problem?) and [gnote](https://gitlab.gnome.org/GNOME/gnote/-/issues/22).

**Comment by herbelin** [comment]
For the record, I could reproduce the bug with the fontselection test of the gtk+ archive and reported it (gtk bug [5226](https://gitlab.gnome.org/GNOME/gtk/-/issues/5226)).

It is about the weight, i.e. boldness, of fonts. There are various conventions and scales (such as multiples of 100 from 100 to 1000, or multiples of 100 plus extra values, or values between 1 and 1000, or values between 0 and 900) and there are conversions to do. At some time, the higher layer of gtk+ finds that a MacOS system font has a weight of 0, which it is not prepared to handle (in _gtk_css_font_weight_value_new, which warns). [I still don't understand though whether this 0 is the trace of a former conversion bug or a correct value in the MacOS collection of fonts.]

**Comment by herbelin** [comment]
One idea to bypass the gtk+ bug is to compile gtk+ with `--enable-debug=yes` mode. In this case, the segfault is replaced by a warning. Could the platform and opam patch the homebrew formula before building?

**Comment by ppedrot** [comment]
@herbelin maybe you can do this in a less violent way, by calling the `GLib.Message.set_fatal_mask` function. IIUC you should be able to locally deactivate the failing `gtk_internal_return_val_if_fail` with it.

**Comment by herbelin** [comment]
Indeed, that would be a less naive direction. On the other side, I'm unsure how to test it. In particular, `gtk_internal_return_val_if_fail` seems to be a static macro.

**Comment by herbelin** [comment]
> The next thing would be to find the root cause and to properly fix it.

The root cause of the problem is now [identified](https://gitlab.gnome.org/GNOME/pango/-/issues/705). It was simply a typo in the pango function converting macos weights to pango/gtk weights. The macos floating number weight rounded to 0 was returned by mistake instead of the integer number pango weight (incidentally, such a typo would have been detected in a language without automatic conversion from floats to integers).

In particular, it is present on all versions of macos that include fonts whose macos weight is outside the range [-0.7, 0.8].

> IMHO the work around by @ppedrot is fine.

Yes, I agree too.

**Comment by MSoegtropIMC** [comment]
@herbelin : very well. I will then do the 2022.09 release with the fix by @ppedrot and expect that we have the real fix for the 2023.03 release.

**Comment by herbelin** [comment]
One more information: the bug was introduced in pango 1.49.4 (November 2021), which explains why this is observed only since recently.

**Comment by Zimmi48** [comment]
Thanks Hugo for the deep investigation and fixing the upstream bug!

In principle, the bug fix should be in pango 1.50.12 (not released yet).

It is my understanding that the workaround that was merged could be reverted as soon as both MacPorts and Homebrew get the fixed pango version (maybe a configure-time check for the pango version could be added instead). [Homebrew](https://formulae.brew.sh/formula/pango#default) seems to be up-to-date with upstream (it currently has pango 1.50.11 that was released just a week ago). [MacPorts](https://ports.macports.org/port/pango/) is lagging behind (last update was 6 months ago), but it [seems easy enough](https://github.com/macports/macports-ports/pull/744) to propose an update.

Should we open an issue to track this for Coq 8.17?
